### PR TITLE
Add URL-synced table state with shareable link support

### PR DESCRIPTION
**Pull request by pixelgrid**
Add URL-synced table state with shareable link support

Summary:
Intent:
- Allow users to share a table view (filters, sorting, search) by copying a URL that restores the exact state when opened
- Provide a foundation for URL-driven table state that can be enabled per-table without affecting existing behaviour

Changes:
- Introduced `url-codecs.ts` with pure serialize/parse functions for globalFilter, columnFilters, and sorting URL params, namespaced by `tableSettingsId` (e.g. `?tb.train/pipelines.gf=eval`)
- Added `useUrlTableState` hook that reads URL params on mount and exposes `buildShareUrl` for generating shareable links
- Extended `useLocalStorageTableState` with opt-in `urlFilters` config; URL state takes priority over localStorage when enabled
- Added `ShareTableUrlButton` component (secondary style, arrowLaunch icon) that copies the shareable URL to clipboard and shows a brief "Copied!" confirmation
- Plumbed `enableShareUrl` through `TableConfig` → `adaptTableConfigToTableProps` → `TableActionBar`
- Enabled the feature on the pipelines list table (`train/pipelines`)
- Datetime filter values (complex objects) are silently excluded from URL encoding to avoid `[object Object]` in the URL

Test Plan:
Ran `npx vitest run --project core` — all existing tests pass, 83 new tests added covering codecs (round-trips, edge cases, security allowlists), the URL state hook, the extended localStorage hook (URL priority layering), and the share button component.

Revert Plan:
Revert this PR via `git revert 99976d40`.

### CODE REVIEW

File: javascript/packages/core/components/table/plugins/state-persistence/__tests__/use-local-storage-table-state.test.tsx [hunk 1]
```diff
@@ -427,4 +450,111 @@ describe('useLocalStorageTableState', () => {
...
   });
+
+  describe('urlFilters', () => {
+    it('returns buildShareUrl as a function when urlFilters is not configured', () => {
+      const { result } = renderHook(
+        () => useLocalStorageTableState({ tableSettingsId: 'test-table' }),
+        buildWrapper([getRouterWrapper()])
+      );
+      expect(typeof result.current.buildShareUrl).toBe('function');
+    });
```
Comment (craigmarker) [inline javascript/packages/core/components/table/plugins/state-persistence/__tests__/use-local-storage-table-state.test.tsx]: What does this test help verify?

File: javascript/packages/core/components/table/plugins/state-persistence/__tests__/use-url-table-state.test.ts [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+      const { result } = renderUrlTableState('?tb.users.co=status,name', {
+        tableSettingsId: 'users',
+        validColumnIds: VALID_IDS,
+        scope: ['columnOrder'],
+      });
+      expect(result.current.urlState?.columnOrder).toEqual(['status', 'name']);
+    });
+
+    it('ignores params for a different tableSettingsId', () => {
+      const { result } = renderUrlTableState('?tb.other.gf=hello');
```
Comment (craigmarker) [inline javascript/packages/core/components/table/plugins/state-persistence/__tests__/use-url-table-state.test.ts]: consider explicitly including a tableSettingsId to clarify what tableSettingsId should be parsed.

File: javascript/packages/core/components/table/plugins/state-persistence/__tests__/use-url-table-state.test.ts [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+      expect(result.current.urlState).toBeNull();
+    });
+
+    it('silently drops invalid column IDs in filters', () => {
+      const { result } = renderUrlTableState('?tb.users.cf=ghost:eq:x,status:eq:active');
```
Comment (craigmarker) [inline javascript/packages/core/components/table/plugins/state-persistence/__tests__/use-url-table-state.test.ts]: same here, consider explicitly including validColumnIDs. or reframe the test names to clarify it is testing _fallbacks_

File: javascript/packages/core/components/table/plugins/state-persistence/__tests__/use-url-table-state.test.ts [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    });
+
+    it('silently drops invalid column IDs in columnOrder', () => {
+      const { result } = renderUrlTableState('?tb.users.co=status,ghost,name', {
+        tableSettingsId: 'users',
+        validColumnIds: VALID_IDS,
+        scope: ['columnOrder'],
+      });
+      expect(result.current.urlState?.columnOrder).toEqual(['status', 'name']);
+    });
```
Comment (craigmarker) [inline javascript/packages/core/components/table/plugins/state-persistence/__tests__/use-url-table-state.test.ts:90]: A disadvantage to abstracting VALID_IDS into a variable versus inlining is it is not clear what valid IDs should be pulled from the table state

```suggestion
    it('silently drops invalid column IDs in columnOrder', () => {
      const { result } = renderUrlTableState('?tb.users.co=status,ghost,name', {
        tableSettingsId: 'users',
        validColumnIds: ['status', 'name', 'some-other-column-id'],
        scope: ['columnOrder'],
      });
      expect(result.current.urlState?.columnOrder).toEqual(['status', 'name']);
    });
```

File: javascript/packages/core/components/table/plugins/state-persistence/__tests__/use-url-table-state.test.ts [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+          scope: ['globalFilter'],
+        }
+      );
+      expect(result.current.urlState?.globalFilter).toBe('hello');
+      expect(result.current.urlState?.sorting).toBeUndefined();
+    });
+  });
+
+  describe('buildShareUrl', () => {
+    it('returns a string URL', () => {
```
Comment (craigmarker) [inline javascript/packages/core/components/table/plugins/state-persistence/__tests__/use-url-table-state.test.ts:119]: what does this test verify?

File: javascript/packages/core/components/table/plugins/state-persistence/__tests__/use-url-table-state.test.ts [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+        sorting: [{ id: 'updatedAt', desc: true }],
+      });
+      expect(url).toContain('tb.users.so=updatedAt%3Adesc');
+    });
+
+    it('replaces existing tb.* params for this tableSettingsId', () => {
+      const { result } = renderUrlTableState('?tb.users.gf=old');
+      const url = result.current.buildShareUrl({ globalFilter: 'new' });
+      expect(url).toContain('tb.users.gf=new');
+      expect(url).not.toContain('old');
```
Comment (craigmarker) [inline javascript/packages/core/components/table/plugins/state-persistence/__tests__/use-url-table-state.test.ts]:
```suggestion
      expect(url).toEqual('tb.users.gf=new');
```

File: javascript/packages/core/components/table/plugins/state-persistence/use-local-storage-table-state.ts [hunk 7]
```diff
@@ -43,69 +57,87 @@ import type {
  *   tableSettingsId: 'projects-table',
  * });
  *
- * // Project-specific filters (recommended for project-scoped data)
```
Comment (craigmarker) [inline javascript/packages/core/components/table/plugins/state-persistence/use-local-storage-table-state.ts:46]: seems this lost a lot of good documentation

File: javascript/packages/core/components/table/plugins/state-persistence/use-local-storage-table-state.ts
Comment (craigmarker) [inline javascript/packages/core/components/table/plugins/state-persistence/use-local-storage-table-state.ts:1]: I have some concern about how this implementation has make this hook pretty complex. I think it could help clarify to have a sibling hook that handles the merging of URL state and local storage state. Or...no need to have local-storage state if all you want is URL-shareability. Right now, you're forced to have local storage & URL-shareability.

**Review by craigmarker — CHANGES_REQUESTED** [review]

File: javascript/packages/core/components/table/components/share-table-url-button/share-table-url-button.tsx [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+  const [css, theme] = useStyletron();
+  const [copied, setCopied] = useState(false);
+
+  const handleClick = () => {
+    const url = buildShareUrl(currentState);
+    void navigator.clipboard.writeText(url).then(() => {
+      setCopied(true);
+      setTimeout(() => setCopied(false), 2000);
+    });
+  };
```
Comment (craigmarker) [inline javascript/packages/core/components/table/components/share-table-url-button/share-table-url-button.tsx:24]: nit: handleClick doesn't tell you what the function does. Consider `copyUrlToClipboard.`

File: javascript/packages/core/components/table/components/table-action-bar/types.ts [hunk 9]
```diff
@@ -53,3 +65,6 @@ export type FilterableColumn<TData extends TableData = TableData> = Omit<
 > &
   ColumnRenderState<TData> &
   Omit<FilteringCapability, 'canFilter'>;
+
+// Re-exported for convenience — callers building share URLs directly can import from here.
+export type { TableState };
```
Comment (craigmarker) [inline javascript/packages/core/components/table/components/table-action-bar/types.ts:70]: What was the reasoning for this export?

File: javascript/packages/core/components/table/table.tsx [hunk 10]
```diff
@@ -29,14 +29,17 @@ import { globalFilterFn } from './utils/global-filter-fn';
...
 
 export function Table<T extends TableData = TableData>(inputProps: TableProps<T>) {
   const props = applyDefaultProps<T>(inputProps);
   const columns = useColumnTransformer<T>(props.columns);
   const [css, theme] = useStyletron();
 
   const { state, initialState } = composeTableState(props.state ?? {});
+  const buildShareUrl = (
+    props.state as InputTableState & { buildShareUrl?: (s: Partial<TableState>) => string }
+  )?.buildShareUrl;
```
Comment (craigmarker) [inline javascript/packages/core/components/table/table.tsx:42]: Why is this `as` cast necessary?

File: javascript/packages/core/components/table/table.tsx [hunk 11]
```diff
@@ -134,6 +137,9 @@ export function Table<T extends TableData = TableData>(inputProps: TableProps<T>
           filterableColumns={transformedColumns.filter(
             (column) => column.canFilter && column.filterMode !== FilterMode.NONE
           )}
+          getShareUrl={
+            buildShareUrl ? () => buildShareUrl(table.getState() as Partial<TableState>) : undefined
```
Comment (craigmarker) [inline javascript/packages/core/components/table/table.tsx:141]: why is this cast necessary?

File: javascript/packages/core/components/table/table.tsx [hunk 12]
```diff
@@ -134,6 +137,9 @@ export function Table<T extends TableData = TableData>(inputProps: TableProps<T>
           filterableColumns={transformedColumns.filter(
             (column) => column.canFilter && column.filterMode !== FilterMode.NONE
           )}
+          getShareUrl={
```
Comment (craigmarker) [inline javascript/packages/core/components/table/table.tsx:140]: can just pass directly, what am I missing?
```suggestion
          getShareUrl={buildShareUrl}
```

File: javascript/packages/core/components/table/plugins/state-persistence/use-url-table-state.ts [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+        const co = raw.co.split(',').filter((id) => validColumnIds.includes(id));
+        if (co.length > 0) state.columnOrder = co;
+      }
+
+      return Object.keys(state).length > 0 ? state : null;
+    },
+    // validColumnIds and scope are intentionally excluded — callers should memoize
+    // these arrays or pass stable references to avoid unnecessary re-runs.
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [location.search, tableSettingsId, paramPrefix]
```
Comment (craigmarker) [inline javascript/packages/core/components/table/plugins/state-persistence/use-url-table-state.ts:61]: I'm confused about this eslint-disabling. Why are this code intentionally excluding properties this memoization relies on?

File: javascript/packages/core/components/table/plugins/state-persistence/use-url-table-state.ts [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+    // Append new state params
+    const newParams = buildTableUrlParams(tableSettingsId, currentState, scope, paramPrefix);
+    for (const [key, value] of newParams.entries()) {
+      url.searchParams.set(key, value);
+    }
+
+    return url.toString();
+  };
+
+  return { urlState, hasUrlState, buildShareUrl };
```
Comment (craigmarker) [inline javascript/packages/core/components/table/plugins/state-persistence/use-url-table-state.ts:96]: I'm not sure memoization in this hook is doing anything, given that the hook returns a fresh buildShareUrl on every render. please demonstrate a use case where the memoization is earned

File: javascript/packages/core/components/table/components/share-table-url-button/share-table-url-button.tsx [hunk 15]
```diff
@@ -0,0 +1,34 @@
+import { useState } from 'react';
```
Comment (craigmarker) [inline javascript/packages/core/components/table/components/share-table-url-button/share-table-url-button.tsx:1]: since this component is only used by table-action-bar, please relocate (and consider renaming to match naming of sibling components) to table-action-bar:

Move to components/table-action-bar/components/table-share-url-button/table-share-url-button.tsx and rename the symbol to TableShareUrlButton.

File: javascript/packages/core/components/table/plugins/state-persistence/url-codecs.ts [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+  'contains',
+] as const);
+
+const MAX_COLUMN_FILTERS = 20;
+const MAX_PARAM_BYTES = 2048;
+
+export const DEFAULT_URL_SCOPE: UrlScope[] = ['globalFilter', 'columnFilters', 'sorting'];
+export const DEFAULT_PARAM_PREFIX = 'tb';
+
+// --- Serialize ---
```
Comment (craigmarker) [inline javascript/packages/core/components/table/plugins/state-persistence/url-codecs.ts:28]: nit: if grouping matters, consider splitting into separate files (eg url-codecs/serialize.ts)

File: javascript/packages/core/components/table/components/share-table-url-button/share-table-url-button.tsx [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+  currentState: Partial<TableState>;
+};
+
+export function ShareTableUrlButton({ buildShareUrl, currentState }: ShareTableUrlButtonProps) {
+  const [css, theme] = useStyletron();
+  const [copied, setCopied] = useState(false);
+
+  const handleClick = () => {
+    const url = buildShareUrl(currentState);
+    void navigator.clipboard.writeText(url).then(() => {
```
Comment (craigmarker) [inline javascript/packages/core/components/table/components/share-table-url-button/share-table-url-button.tsx:20]: what happens on failure?

File: javascript/packages/core/components/table/plugins/state-persistence/use-url-table-state.ts [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [location.search, tableSettingsId, paramPrefix]
+  );
+
+  const hasUrlState = urlState !== null;
+
+  const buildShareUrl = (currentState: Partial<TableState>): string => {
+    // Use window.location.origin + router location as the base so that both
+    // production (BrowserRouter) and test (MemoryRouter) environments work correctly.
+    const base = typeof window !== 'undefined' ? window.location.origin : 'http://localhost:3000';
```
Comment (craigmarker) [inline javascript/packages/core/components/table/plugins/state-persistence/use-url-table-state.ts:69]: I don't think we should be falling back to localhost:3000, since this is a shared library that might have deployments with different hostnames.

File: javascript/packages/core/components/table/plugins/state-persistence/__tests__/use-table-state.test.ts [hunk 19]
```diff
@@ -0,0 +1,198 @@
...
+describe('useTableState', () => {
+  beforeEach(() => {
+    localStorage.clear();
+    vi.clearAllMocks();
+  });
+
+  describe('without urlFilters', () => {
+    const VALID_COLUMN_IDS = ['name', 'department', 'status'];
+
+    it('always exposes buildShareUrl as a callable function', () => {
```
Comment (craigmarker) [inline javascript/packages/core/components/table/plugins/state-persistence/__tests__/use-table-state.test.ts:18]: can this test what happens when buildShareUrl is called instead of the fact that it is callable (which is an implementation detail)

File: javascript/packages/core/components/table/plugins/state-persistence/__tests__/use-table-state.test.ts [hunk 20]
```diff
@@ -0,0 +1,198 @@
...
+            validColumnIds: VALID_COLUMN_IDS,
+            urlFilters: { enabled: true },
+          }),
+        buildWrapper([getRouterWrapper()])
+      );
+
+      expect(result.current.globalFilter).toBe('fromlocal');
+    });
+
+    it('silently drops invalid column IDs from URL params', () => {
```
Comment (craigmarker) [inline javascript/packages/core/components/table/plugins/state-persistence/__tests__/use-table-state.test.ts:113]: this test becomes more self-documenting is VALID_COLUMN_IDS are inlined

File: javascript/packages/core/components/table/plugins/state-persistence/__tests__/use-table-state.test.ts [hunk 21]
```diff
@@ -0,0 +1,198 @@
...
+      >;
+      expect(stored['test-table'].globalFilter).toBe('hello');
+    });
+
+    it('ignores URL params when urlFilters is not enabled', () => {
+      const { result } = renderHook(
+        () =>
+          useTableState({
+            tableSettingsId: 'test-table',
+            validColumnIds: VALID_COLUMN_IDS,
```
Comment (craigmarker) [inline javascript/packages/core/components/table/plugins/state-persistence/__tests__/use-table-state.test.ts:49]: what does the inclusion of validColumnIds add to this test case?

**Review by craigmarker — CHANGES_REQUESTED** [review]
